test/parallel/test-fs-rm.js

```diff
@@ -287,24 +286,51 @@ function removeAsync(dir) {
   // IBMi has a different access permission mechanism
   // This test should not be run as `root`
   if (!common.isIBMi && (common.isWindows || process.getuid() !== 0)) {
-    function makeDirectoryReadOnly(dir, mode) {
-      let accessErrorCode = 'EACCES';
-      if (common.isWindows) {
-        accessErrorCode = 'EPERM';
-        execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(DE,DC)"`);
-      } else {
-        fs.chmodSync(dir, mode);
+    // On Windows, we are allowed to access and modify the contents of a
+    // read-only folder.
+    {
+      // Check that deleting a file that cannot be accessed using rmsync throws
+      // https://github.com/nodejs/node/issues/38683
+      const dirname = nextDirPath();
+      const filePath = path.join(dirname, 'text.txt');
+
+      fs.mkdirSync(dirname, { recursive: true });
+      fs.writeFileSync(filePath, 'hello');
+
+      fs.chmodSync(filePath, 0o444);
+      fs.chmodSync(dirname, 0o444);
+
+      let err = null;
+
+      try {
+        fs.rmSync(filePath, { force: true });
+      } catch (_err) {
+        err = _err;
       }
-      return accessErrorCode;
-    }
 
-    function makeDirectoryWritable(dir) {
-      if (fs.existsSync(dir)) {
-        if (common.isWindows) {
-          execSync(`icacls ${dir} /remove:d "everyone"`);
-        } else {
-          fs.chmodSync(dir, 0o777);
-        }
+      try {
+        fs.chmodSync(dirname, 0o777);
+      } catch {
+      }
+
+      try {
+        fs.chmodSync(filePath, 0o777);
+      } catch {
+      }
+
+      let isValidState = true;
+      const exists = fs.existsSync(filePath);
+
+      if (common.isWindows &&
+          !(exists === false && err === null)) {
+        isValidState = false;
+      } else if (!common.isWindows &&
+                 !(exists === true && err?.code === 'EACCES')) {
+        isValidState = false;
+      }
```

Comment:
```suggestion
      const exists = fs.existsSync(filePath);
      const isValidState = common.isWindows ? (exists || err != null) : (!exist || err?.code !== 'EACCES');
```

Reply:
Here, `isValidState` is supposed to store the negation of the result of the ternary. TBH, I did initially consider using a ternary here and I also thought of placing the expression directly in the conditional where `err` is thrown but it felt a little unreadable to me.

Other than separating out the conditions into the Windows and the non-Windows cases, I used negations because that way, I can enclose what to expect within `!()` instead of writing out the unexpected possibilities.

I can change this if you feel strongly about it. :)

Reply:
To me the ternary is more readable, but I don't feel strongly about it.

> I used negations because that way, I can enclose what to expect within !() instead of writing out the unexpected possibilities.

I'm probably stating the obvious, but you can negate the expected condition with a ternary:
```suggestion
      const exists = fs.existsSync(filePath);
      const isValidState = common.isWindows ? !(!exists && err == null) : !(exist && err?.code === 'EACCES');
```
I'd still prefer my original suggestion, but no strong feeling either :)

In any case, I think that whole section should be moved to a function to avoid the code duplication.

Reply:
@aduh95 I moved it into a ternary with some indentation to make it easier to understand. :)